File: scripts/read_geqdsk.py
#!/usr/bin/env python
import numpy as np
    
def read_geqdsk(filename):
    """
    'desc'    :desc,  
    'nw'      :nw,    # Numver of horizontal R grid  points
    'nh'      :nh,    # Number of vertical Z grid points
    'rdim'    :rdim,  # Horizontal dimension in meter of computational box
    'zdim'    :zdim,  # Vertical dimension in meter of computational box
    'rcentr'  :rcentr,# 
    'rleft'   :rleft, # Minimum R in meter of rectangular computational box
    'zmid'    :zmid,  # Z of center of computational box in meter
    'rmaxis'  :rmaxis,# R of magnetic axis in meter
    'rmaxis'  :zmaxis,# Z of magnetic axis in meter
    'simag'   :simag, # poloidal flus ax magnetic axis in Weber / rad
    'sibry'   :sibry, # Poloidal flux at the plasma boundary in Weber / rad
    'rcentr'  :rcentr,# R in meter of  vacuum toroidal magnetic field BCENTR
    'bcentr'  :bcentr,# Vacuum toroidal magnetic field in Tesla at RCENTR
    'current' :current,# Plasma current in Ampere 
    'fpol'    :fpol,  # Poloidal current function in m-T, $F=RB_T$ on flux grid
    'pres'    :pres,  # Plasma pressure in $nt/m^2$ on uniform flux grid
    'ffprim'  :ffprim,# $FF^\prime(\psi)$ in $(mT)^2/(Weber/rad)$ on uniform flux grid
    'pprim'   :pprim, # $P^\prime(\psi)$ in $(nt/m^2)/(Weber/rad)$ on uniform flux grid
    'psizr'   :psizr, # Poloidal flus in Webber/rad on the rectangular grid points
    'qpsi'    :qpsi,  # q values on uniform flux grid from axis to boundary
    'nbbbs'   :nbbbs, # Number of boundary points
    'limitr'  :limitr,# Number of limiter points
    'rbbbs'   :rbbbs, # R of boundary points in meter
    'zbbbs'   :zbbbs, # Z of boundary points in meter
    'rlim'    :rlim,  # R of surrounding limiter contour in meter
    'rlim'    :zlim,  # R of surrounding limiter contour in meter
    
    Toroidal Current Density
     $J_T(Amp/m^2)= R P^\prim(\psi)+ F F^\prim(\psi)/R/\mu_0$
    """
    d=open(filename,"r").read().replace("\n","");
    desc	= d[0:48]
   # idum	= int(d[48:52])
    nw	     = int(d[52:56])
    nh	     = int(d[56:60])
    it=60
    (rdim,zdim,rcentr,rleft,zmid,
    rmaxis,zmaxis,simag,sibry,bcentr,
    current,simag,xdum,rmaxis,xdum,
    zmaxis,xdum,sibry,xdum,xdum)=(float(d[it+i*16:it+(i+1)*16]) for i in range(20))
    it+=20*16;
    
    fpol=np.array([float(d[it+i*16:it+(i+1)*16]) for i in range(nw)]);
    it+=nw*16;
    
    pres=np.array([float(d[it+i*16:it+(i+1)*16]) for i in range(nw)]);
    it+=nw*16;
    
    ffprim=np.array([float(d[it+i*16:it+(i+1)*16]) for i in range(nw)]);
    it+=nw*16;
    
    pprime=np.array([float(d[it+i*16:it+(i+1)*16]) for i in range(nw)]);
    it+=nw*16;
    
    psirz=np.reshape(np.array([float(d[it+i*16:it+(i+1)*16]) for i in range(nw*nh)]),(nw,nh));
    it+=nh*nw*16;
    
    qpsi=np.array([float(d[it+i*16:it+(i+1)*16]) for i in range(nw)]);
    it+=nw*16;
    
    nbbbs=int(d[it:it+5])
    limitr=int(d[it+5:it+10])
    it+=10
    
    rbbbs=np.array([float(d[it+i*32:it+i*32+16]) for i in range(nbbbs)]);
    zbbbs=np.array([float(d[it+i*32+16:it+i*32+32]) for i in range(nbbbs)]);
    it+=nbbbs*16*2;
    
    rlim=np.array([float(d[it+i*32:it+i*32+16]) for i in range(limitr)]);
    zlim=np.array([float(d[it+i*32+16:it+i*32+32]) for i in range(limitr)]);
    
    
    return {
    'desc'    :desc,  
    'nw'      :nw,    # Number of horizontal R grid  points
    'nh'      :nh,    # Number of vertical Z grid points
    'rdim'    :rdim,  # Horizontal dimension in meter of computational box
    'zdim'    :zdim,  # Vertical dimension in meter of computational box
    'rcentr'  :rcentr,# 
    'rleft'   :rleft, # Minimum R in meter of rectangular computational box
    'zmid'    :zmid,  # Z of center of computational box in meter
    'rmaxis'  :rmaxis,# R of magnetic axis in meter
    'zmaxis'  :zmaxis,# Z of magnetic axis in meter
    'simag'   :simag, # poloidal flus ax magnetic axis in Weber / rad
    'sibry'   :sibry, # Poloidal flux at the plasma boundary in Weber / rad
    'rcentr'  :rcentr,# R in meter of  vacuum toroidal magnetic field BCENTR
    'bcentr'  :bcentr,# Vacuum toroidal magnetic field in Tesla at RCENTR 
    'current' :current,# Plasma current in Ampere
    'fpol'    :fpol,  # Poloidal current function in m-T, $F=RB_T$ on flux grid
    'pres'    :pres,  # Plasma pressure in $nt/m^2$ on uniform flux grid
    'ffprim'  :ffprim,# $FF^\prime(\psi)$ in $(mT)^2/(Weber/rad)$ on uniform flux grid
    'pprime'  :pprime,# $P^\prime(\psi)$ in $(nt/m^2)/(Weber/rad)$ on uniform flux grid
    'psizr'   :psirz, # Poloidal flus in Webber/rad on the rectangular grid points
    'qpsi'    :qpsi,  # q values on uniform flux grid from axis to boundary
    'nbbbs'   :nbbbs, # Number of boundary points
    'limitr'  :limitr,# Number of limiter points
    'rbbbs'   :rbbbs, # R of boundary points in meter
    'zbbbs'   :zbbbs, # Z of boundary points in meter
    'rlim'    :rlim,  # R of surrounding limiter contour in meter
    'rlim'    :zlim,  # R of surrounding limiter contour in meter
    }
def geqdsk2gmsh(data,filename):
    f=open(filename,"w");
    
        
    return
#     plt.contour(psirz.reshape([nh,nw]),[(sibry-simag)/10.0*i+simag for i in range(10)])
#     
#     plt.plot((rlims-rleft)/rdim*nw,zlims/zdim*nh+nh/2)
#     
#     plt.plot((rbbbs-rleft)/rdim*nw,zbbbs/zdim*nh+nh/2)
#     
#     plt.show()

File: src/utilities/read_geqdsk.cpp
/*
 * read_geqdsk.cpp
 *
 *  Created on: 2013年11月29日
 *      Author: salmon
 */

#include "read_geqdsk.h"
#include "pertty_ostream.h"
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <iterator>
//#include <memory>
#include <string>
#include <vector>

namespace simpla
{

void GEqdsk::Read()
{
	inFileStream_.get(desc, 48);

	int idum;

	inFileStream_ >> std::setw(4) >> idum >> nw >> nh;

	double xdum;

	inFileStream_ >> std::setw(16)

	>> rdim >> zdim >> rcentr >> rleft >> zmid

	>> rmaxis >> zmaxis >> simag >> sibry >> bcentr

	>> current >> simag >> xdum >> rmaxis >> xdum

	>> zmaxis >> xdum >> sibry >> xdum >> xdum;

	fpol.resize(nw);
	pres.resize(nw);
	ffprim.resize(nw);
	pprim.resize(nw);
	qpsi.resize(nw);
	psirz.resize(nw * nh);

	inFileStream_

	>> std::setw(16) >> fpol

	>> std::setw(16) >> pres

	>> std::setw(16) >> ffprim

	>> std::setw(16) >> pprim

	>> std::setw(16) >> psirz

	>> std::setw(16) >> qpsi;

	inFileStream_ >> std::setw(5) >> nbbbs >> limitr;

	rzbbb.resize(nbbbs * 2);
	rzlim.resize(limitr * 2);

	inFileStream_ >> std::setw(16) >> rzbbb >> rzlim;
}
std::ostream & GEqdsk::Print(std::ostream & os)
{
	std::cout << "--" << desc << std::endl;

	std::cout << "nw" << "\t= " << nw
			<< "\t--  Number of horizontal R grid  points" << std::endl;

	std::cout << "nh" << "\t= " << nh << "\t-- Number of vertical Z grid points"
			<< std::endl;

	std::cout << "rdim" << "\t= " << rdim
			<< "\t-- Horizontal dimension in meter of computational box                   "
			<< std::endl;

	std::cout << "zdim" << "\t= " << zdim
			<< "\t-- Vertical dimension in meter of computational box                   "
			<< std::endl;

	std::cout << "rcentr" << "\t= " << rcentr
			<< "\t--                                                                    "
			<< std::endl;

	std::cout << "rleft" << "\t= " << rleft
			<< "\t-- Minimum R in meter of rectangular computational box                "
			<< std::endl;

	std::cout << "zmid" << "\t= " << zmid
			<< "\t-- Z of center of computational box in meter                          "
			<< std::endl;

	std::cout << "rmaxis" << "\t= " << rmaxis
			<< "\t-- R of magnetic axis in meter                                        "
			<< std::endl;

	std::cout << "rmaxis" << "\t= " << zmaxis
			<< "\t-- Z of magnetic axis in meter                                        "
			<< std::endl;

	std::cout << "simag" << "\t= " << simag
			<< "\t-- poloidal flus ax magnetic axis in Weber / rad                      "
			<< std::endl;

	std::cout << "sibry" << "\t= " << sibry
			<< "\t-- Poloidal flux at the plasma boundary in Weber / rad                "
			<< std::endl;

	std::cout << "rcentr" << "\t= " << rcentr
			<< "\t-- R in meter of  vacuum toroidal magnetic field BCENTR               "
			<< std::endl;

	std::cout << "bcentr" << "\t= " << bcentr
			<< "\t-- Vacuum toroidal magnetic field in Tesla at RCENTR                  "
			<< std::endl;

	std::cout << "current" << "\t= " << current
			<< "\t-- Plasma current in Ampere                                          "
			<< std::endl;

	std::cout << "fpol" << "\t= "
			<< "\t-- Poloidal current function in m-T<< $F=RB_T$ on flux grid           "
			<< std::endl << fpol << std::endl;

	std::cout << "pres" << "\t= "
			<< "\t-- Plasma pressure in $nt/m^2$ on uniform flux grid                   "
			<< std::endl << pres << std::endl;

	std::cout << "ffprim" << "\t= "
			<< "\t-- $FF^\\prime(\\psi)$ in $(mT)^2/(Weber/rad)$ on uniform flux grid     "
			<< std::endl << ffprim << std::endl;

	std::cout << "pprim" << "\t= "
			<< "\t-- $P^\\prime(\\psi)$ in $(nt/m^2)/(Weber/rad)$ on uniform flux grid    "
			<< std::endl << pprim << std::endl;

	std::cout << "psizr"
			<< "\t-- Poloidal flus in Webber/rad on the rectangular grid points         "
			<< std::endl << psirz << std::endl;

	std::cout << "qpsi" << "\t= "
			<< "\t-- q values on uniform flux grid from axis to boundary                "
			<< std::endl << qpsi << std::endl;

	std::cout << "nbbbs" << "\t= " << nbbbs
			<< "\t-- Number of boundary points                                          "
			<< std::endl;

	std::cout << "limitr" << "\t= " << limitr
			<< "\t-- Number of limiter points                                           "
			<< std::endl;

	std::cout << "rzbbbs" << "\t= "
			<< "\t-- R of boundary points in meter                                      "
			<< std::endl << rzbbb << std::endl;

	std::cout << "rzlim" << "\t= "
			<< "\t-- R of surrounding limiter contour in meter                          "
			<< std::endl << rzlim << std::endl;

	return os;
}
}  // namespace simpla
using namespace simpla;
int main(int argc, char ** argv)
{
	GEqdsk geqdsk(argv[1]);

	geqdsk.Read();
	geqdsk.Print(std::cout);
}
